### PR TITLE
[ENHANCE] Integration of recipe and model configuration

### PR DESCRIPTION
**Pull request by harimkang**
## Summary
As part of a refactoring to reduce unnecessary patching, incorporate `recipe.yaml` into model.py so that it is also visible in workspace `model.py`. This may be fixed in the future.
Import the recipe file as _base_ in model.py.
Unify the Task's `_model_cfg` with the `_recipe_cfg`.
Remove code for loading recipe file

## Changes
model.py == model.py + recipe.yaml
image

Do not merge until the test passes.

### CODE REVIEW

File: otx/algorithms/common/tasks/nncf_base.py [hunk 1]
```diff
@@ -164,7 +164,7 @@ def _init_nncf_cfg(self):
 
     def _initialize_post_hook(self, options=None):
         super()._initialize_post_hook(options)
-        assert self._recipe_cfg is not None and self._model_cfg is not None
+        assert self._recipe_cfg is not None and self._recipe_cfg is not None
```
Comment (eunwoosh) [inline otx/algorithms/common/tasks/nncf_base.py]:
```suggestion
        assert self._recipe_cfg is not None
```

File: otx/algorithms/common/tasks/nncf_base.py [hunk 2]
```diff
@@ -181,9 +181,8 @@ def _initialize_post_hook(self, options=None):
                 self._recipe_cfg.data["train_dataloader"] = data_loader
 
         # nncf does not suppoer FP16
-        if "fp16" in self._recipe_cfg or "fp16" in self._model_cfg:
+        if "fp16" in self._recipe_cfg or "fp16" in self._recipe_cfg:
```
Comment (eunwoosh) [inline otx/algorithms/common/tasks/nncf_base.py]:
```suggestion
        if "fp16" in self._recipe_cfg:
```

File: otx/algorithms/classification/tasks/inference.py [hunk 3]
```diff
@@ -440,32 +438,9 @@ def _init_recipe_hparam(self) -> dict:
...
-                self._model_dir = os.path.join(self._model_dir, "supcon")
-            else:
-                recipe = os.path.join(recipe_root, RECIPE_TRAIN_TYPE[self._train_type])
-        else:
-            raise NotImplementedError(f"Train type {self._train_type} is not implemented yet.")
-
-        logger.info(f"train type = {self._train_type} - loading {recipe}")
-
-        self._recipe_cfg = MPAConfig.fromfile(recipe)
+        self._recipe_cfg = self._init_model_cfg()
```
Comment (eunwoosh) [inline otx/algorithms/classification/tasks/inference.py:454]: How about changing function name `_init_model_cfg` to `_init_recipe` then?

File: otx/algorithms/common/tasks/training_base.py [hunk 4]
```diff
@@ -136,16 +135,16 @@ def _run_task(self, stage_module, mode=None, dataset=None, **kwargs):
...
-        assert model_cfg is not None, "'model_cfg' is not initialized."
+        # assert model_cfg is not None, "'model_cfg' is not initialized."
         assert recipe_cfg is not None, "'recipe_cfg' is not initialized."
 
         # update model config -> model label schema
         data_classes = [label.name for label in self._labels]
         model_classes = [label.name for label in self._model_label_schema]
-        model_cfg["model_classes"] = model_classes
+        recipe_cfg["model_classes"] = model_classes
+        # model_cfg["model_classes"] = model_classes
```
Comment (eunwoosh) [inline otx/algorithms/common/tasks/training_base.py]: how about deleting it

File: otx/algorithms/common/tasks/training_base.py [hunk 5]
```diff
@@ -136,16 +135,16 @@ def _run_task(self, stage_module, mode=None, dataset=None, **kwargs):
 
         # deepcopy all configs to make sure
         # changes under MPA and below does not take an effect to OTX for clear distinction
-        model_cfg = deepcopy(self._model_cfg)
         recipe_cfg = deepcopy(self._recipe_cfg)
         data_cfg = deepcopy(self._data_cfg)
-        assert model_cfg is not None, "'model_cfg' is not initialized."
+        # assert model_cfg is not None, "'model_cfg' is not initialized."
```
Comment (eunwoosh) [inline otx/algorithms/common/tasks/training_base.py]: how about deleting it

File: otx/algorithms/classification/tasks/inference.py (hunk 3 above)
Comment (harimkang) [reply otx/algorithms/classification/tasks/inference.py:454]: Please fix this in refactoring, it's too big a scope to cover in this PR.

File: otx/algorithms/common/tasks/nncf_base.py (hunk 1 above)
Comment (harimkang) [reply otx/algorithms/common/tasks/nncf_base.py]: Thanks, Done!

File: otx/algorithms/common/tasks/nncf_base.py (hunk 2 above)
Comment (harimkang) [reply otx/algorithms/common/tasks/nncf_base.py]: Thanks, Done!

File: otx/algorithms/common/tasks/training_base.py (hunk 4 above)
Comment (harimkang) [reply otx/algorithms/common/tasks/training_base.py]: Thanks, Done!

File: otx/algorithms/common/tasks/training_base.py (hunk 5 above)
Comment (harimkang) [reply otx/algorithms/common/tasks/training_base.py]: Thanks, Done!

**Comment by harimkang** [comment]
@eunwoosh  Thank you for your review. This PR is still WIP, so I'll request a review again later.

**Comment by JihwanEom** [comment]
@sovrasov Could you take a look about https://github.com/openvinotoolkit/training_extensions/pull/1875/files#diff-92be97451ac7ffd50f8ee797e8d52f85aeeaacf3099700799ffbc68286671e84R45? `cfg.merge_from_dict(model_cfg._cfg_dict)`:https://github.com/openvinotoolkit/training_extensions/blob/develop/otx/mpa/cls/stage.py#L37 is assuming only string key, not integer key so below error was caused. 
I updated to output string key:val in `head_idx_to_logits_range`. Can you share your concern about this change?

```
E       AssertionError: Traceback (most recent call last):
E         File "/home/cosmos/anaconda3/envs/otx-dev/bin/otx", line 8, in <module>
E           sys.exit(main())
E         File "/home/cosmos/ws/otx-dev/otx/cli/tools/cli.py", line 76, in main
E           results = globals()[f"otx_{name}"]()
E         File "/home/cosmos/ws/otx-dev/otx/cli/tools/train.py", line 219, in main
E           task.train(dataset, output_model, train_parameters=TrainParameters())
E         File "/home/cosmos/ws/otx-dev/otx/api/utils/argument_checks.py", line 234, in validate
E           return function(**input_parameters_values_map)
E         File "/home/cosmos/ws/otx-dev/otx/algorithms/classification/tasks/train.py", line 119, in train
E           results = self._run_task(stage_module, mode="train", dataset=dataset, parameters=train_parameters)
E         File "/home/cosmos/ws/otx-dev/otx/algorithms/common/tasks/training_base.py", line 168, in _run_task
E           output = workflow.run(
E         File "/home/cosmos/ws/otx-dev/otx/mpa/cls/trainer.py", line 38, in run
E           cfg = self.configure(model_cfg, model_ckpt, data_cfg, training=True, **kwargs)
E         File "/home/cosmos/ws/otx-dev/otx/mpa/cls/stage.py", line 28, in configure
E           self.configure_model(cfg, model_cfg, training, **kwargs)
E         File "/home/cosmos/ws/otx-dev/otx/mpa/cls/stage.py", line 37, in configure_model
E           cfg.merge_from_dict(model_cfg._cfg_dict)
E         File "/home/cosmos/anaconda3/envs/otx-dev/lib/python3.8/site-packages/mmcv/utils/config.py", line 644, in merge_from_dict
E           Config._merge_a_into_b(
E         File "/home/cosmos/anaconda3/envs/otx-dev/lib/python3.8/site-packages/mmcv/utils/config.py", line 327, in _merge_a_into_b
E           b[k] = Config._merge_a_into_b(v, b[k], allow_list_keys)
E         File "/home/cosmos/anaconda3/envs/otx-dev/lib/python3.8/site-packages/mmcv/utils/config.py", line 327, in _merge_a_into_b
E           b[k] = Config._merge_a_into_b(v, b[k], allow_list_keys)
E         File "/home/cosmos/anaconda3/envs/otx-dev/lib/python3.8/site-packages/mmcv/utils/config.py", line 327, in _merge_a_into_b
E           b[k] = Config._merge_a_into_b(v, b[k], allow_list_keys)
E         [Previous line repeated 1 more time]
E         File "/home/cosmos/anaconda3/envs/otx-dev/lib/python3.8/site-packages/mmcv/utils/config.py", line 312, in _merge_a_into_b
E           if allow_list_keys and k.isdigit() and isinstance(b, list):
E       AttributeError: 'int' object has no attribute 'isdigit'
```

**Comment by harimkang** [comment]
I merged the recipe configuration into model.py.  Please check if the existing configuration works as it is.

**Comment by harimkang** [comment]
@sungchul2 Could you review the `model.py` on the selfsl and supcon side to make sure the changes are correct?
@supersoob @JihwanEom @jaegukhyun Could you review each task's `model.py` changes once?

File: otx/algorithms/classification/configs/efficientnet_b0_cls_incr/selfsl/model.py [hunk 6]
```diff
@@ -2,16 +2,23 @@
...
+    ),
     base_momentum=0.996,
     neck=dict(type="SelfSLMLP", in_channels=1280, hid_channels=4096, out_channels=256, with_avg_pool=True),
     head=dict(
         type="ConstrastiveHead",
         predictor=dict(type="SelfSLMLP", in_channels=256, hid_channels=4096, out_channels=256, with_avg_pool=False),
+        loss=dict(
+            type="CrossEntropyLoss",
+            loss_weight=1.0,
+        ),
```
Comment (sungchul1) [inline otx/algorithms/classification/configs/efficientnet_b0_cls_incr/selfsl/model.py]: You can remove this loss arg in self-sl head because loss will be automatically calculated through BYOL loss in the head.

File: otx/algorithms/segmentation/configs/ocr_lite_hrnet_18/semisl/model.py [hunk 7]
```diff
@@ -17,6 +17,44 @@
 # pylint: disable=invalid-name
 
 _base_ = [
-    "../../base/models/mean_teacher.py",
```
Comment (supersoob) [inline otx/algorithms/segmentation/configs/ocr_lite_hrnet_18/semisl/model.py:20]: I suggest to remove this file if all model config is specified here!

File: otx/algorithms/classification/configs/efficientnet_v2_s_cls_incr/selfsl/model.py [hunk 8]
```diff
@@ -2,16 +2,20 @@
...
     task="classification",
+    backbone=dict(
+        version="s_21k",
+    ),
     base_momentum=0.996,
     neck=dict(type="SelfSLMLP", in_channels=1280, hid_channels=4096, out_channels=256, with_avg_pool=True),
     head=dict(
         type="ConstrastiveHead",
         predictor=dict(type="SelfSLMLP", in_channels=256, hid_channels=4096, out_channels=256, with_avg_pool=False),
+        loss=dict(type="CrossEntropyLoss", loss_weight=1.0),
```
Comment (sungchul1) [inline otx/algorithms/classification/configs/efficientnet_v2_s_cls_incr/selfsl/model.py]: Like above.

**Review by jaegukhyun — DISMISSED** [review]
For detection part, there is no big objection

File: otx/algorithms/classification/configs/mobilenet_v3_large_075_cls_incr/selfsl/model.py [hunk 9]
```diff
@@ -2,16 +2,23 @@
...
+        mode="large",
+        width_mult=0.75,
+    ),
     base_momentum=0.996,
     neck=dict(type="SelfSLMLP", in_channels=720, hid_channels=4096, out_channels=256, with_avg_pool=True),
     head=dict(
         type="ConstrastiveHead",
         predictor=dict(type="SelfSLMLP", in_channels=256, hid_channels=4096, out_channels=256, with_avg_pool=False),
+        in_channels=720,
+        hid_channels=1280,
```
Comment (sungchul1) [inline otx/algorithms/classification/configs/mobilenet_v3_large_075_cls_incr/selfsl/model.py]: These args are not needed because they are already assigned at L19.

File: otx/algorithms/classification/configs/mobilenet_v3_large_1_cls_incr/selfsl/model.py [hunk 10]
```diff
@@ -2,16 +2,23 @@
...
     neck=dict(type="SelfSLMLP", in_channels=960, hid_channels=4096, out_channels=256, with_avg_pool=True),
     head=dict(
         type="ConstrastiveHead",
         predictor=dict(type="SelfSLMLP", in_channels=256, hid_channels=4096, out_channels=256, with_avg_pool=False),
+        in_channels=960,
+        hid_channels=1280,
+        loss=dict(
+            type="CrossEntropyLoss",
+            loss_weight=1.0,
+        ),
```
Comment (sungchul1) [inline otx/algorithms/classification/configs/mobilenet_v3_large_1_cls_incr/selfsl/model.py]: Like above.

**Review by supersoob — DISMISSED** [review]
Overall, it looks good. Thanks!

File: otx/algorithms/classification/configs/efficientnet_b0_cls_incr/selfsl/model.py (hunk 6 above)
Comment (sungchul1) [reply otx/algorithms/classification/configs/efficientnet_b0_cls_incr/selfsl/model.py]: And it looks better to insert `_delete_=True` in `head` not to be inherited anything from the previous one.
(It is not essential, but just for making the recipe more clear.)

File: otx/algorithms/segmentation/configs/ocr_lite_hrnet_18/semisl/model.py (hunk 7 above)
Comment (harimkang) [reply otx/algorithms/segmentation/configs/ocr_lite_hrnet_18/semisl/model.py:20]: Thank you. Deleted in this [commit](https://github.com/openvinotoolkit/training_extensions/pull/1875/commits/13fa9694d5b23702bf064005474f533fab16b559).

File: otx/algorithms/segmentation/configs/ocr_lite_hrnet_18/selfsl/model.py [hunk 11]
```diff
@@ -46,6 +48,29 @@
...
+        enable_aggregator=True,
+        enable_out_norm=False,
+        loss_decode=[
+            dict(
+                type="CrossEntropyLoss",
+                use_sigmoid=False,
+                loss_weight=1.0,
+            ),
+        ],
+    ),
```
Comment (sungchul1) [inline otx/algorithms/segmentation/configs/ocr_lite_hrnet_18/selfsl/model.py]: This `decode_head` args is not needed in this recipe like above.

File: otx/algorithms/segmentation/configs/ocr_lite_hrnet_18_mod2/selfsl/model.py [hunk 12]
```diff
@@ -46,6 +48,29 @@
...
+        enable_aggregator=True,
+        enable_out_norm=False,
+        loss_decode=[
+            dict(
+                type="CrossEntropyLoss",
+                use_sigmoid=False,
+                loss_weight=1.0,
+            ),
+        ],
+    ),
```
Comment (sungchul1) [inline otx/algorithms/segmentation/configs/ocr_lite_hrnet_18_mod2/selfsl/model.py]: Like above.

File: otx/algorithms/segmentation/configs/ocr_lite_hrnet_s_mod2/selfsl/model.py [hunk 13]
```diff
@@ -46,6 +47,38 @@
...
+                loss_weight=1.0,
+            ),
+        ],
+        init_cfg=dict(
+            type="Normal",
+            mean=0,
+            std=0.01,
+            override=dict(name="conv_seg"),
+        ),
+    ),
```
Comment (sungchul1) [inline otx/algorithms/segmentation/configs/ocr_lite_hrnet_s_mod2/selfsl/model.py]: Like above.

File: otx/algorithms/segmentation/configs/ocr_lite_hrnet_x_mod3/selfsl/model.py [hunk 14]
```diff
@@ -46,6 +47,33 @@
...
+        enable_out_norm=False,
+        enable_loss_equalizer=True,
+        loss_decode=[
+            dict(
+                type="CrossEntropyLoss",
+                use_sigmoid=False,
+                loss_weight=1.0,
+            ),
+        ],
+    ),
```
Comment (sungchul1) [inline otx/algorithms/segmentation/configs/ocr_lite_hrnet_x_mod3/selfsl/model.py]: Like above.

**Review by jaegukhyun — DISMISSED** [review]

**Review by sungchul1 — CHANGES_REQUESTED** [review]
Thanks for your hard work! I left some comments.

File: otx/algorithms/classification/configs/efficientnet_b0_cls_incr/selfsl/model.py (hunk 6 above)
Comment (harimkang) [reply otx/algorithms/classification/configs/efficientnet_b0_cls_incr/selfsl/model.py]: Thanks for the review. I've fixed it, but can you double check? https://github.com/openvinotoolkit/training_extensions/pull/1875/commits/420fb5b81daa29b4237f5f8cc6de1bf874ddf8ae

**Review by sungchul1 — APPROVED** [review]

**Review by JihwanEom — APPROVED** [review]

File: otx/algorithms/classification/adapters/mmcls/data/datasets.py [hunk 15]
```diff
@@ -387,9 +387,9 @@ def evaluate(
         for i in range(self.hierarchical_info["num_multiclass_heads"]):
             multiclass_logit = results[
                 :,
-                self.hierarchical_info["head_idx_to_logits_range"][i][0] : self.hierarchical_info[
+                self.hierarchical_info["head_idx_to_logits_range"][str(i)][0] : self.hierarchical_info[
```
Comment (sovrasov) [inline otx/algorithms/classification/adapters/mmcls/data/datasets.py:390]: Why this change is required? The data type is supposed to be integer here. Is this change related to the other ones in this PR? Previously, we had no any issues with this code, as far, as I know.

**Comment by sovrasov** [comment]
> @sovrasov Could you take a look about https://github.com/openvinotoolkit/training_extensions/pull/1875/files#diff-92be97451ac7ffd50f8ee797e8d52f85aeeaacf3099700799ffbc68286671e84R45? `cfg.merge_from_dict(model_cfg._cfg_dict)`:https://github.com/openvinotoolkit/training_extensions/blob/develop/otx/mpa/cls/stage.py#L37 is assuming only string key, not integer key so below error was caused. I updated to output string key:val in `head_idx_to_logits_range`. Can you share your concern about this change?
> 
> ```
> E       AssertionError: Traceback (most recent call last):
> E         File "/home/cosmos/anaconda3/envs/otx-dev/bin/otx", line 8, in <module>
> E           sys.exit(main())
> E         File "/home/cosmos/ws/otx-dev/otx/cli/tools/cli.py", line 76, in main
> E           results = globals()[f"otx_{name}"]()
> E         File "/home/cosmos/ws/otx-dev/otx/cli/tools/train.py", line 219, in main
> E           task.train(dataset, output_model, train_parameters=TrainParameters())
> E         File "/home/cosmos/ws/otx-dev/otx/api/utils/argument_checks.py", line 234, in validate
> E           return function(**input_parameters_values_map)
> E         File "/home/cosmos/ws/otx-dev/otx/algorithms/classification/tasks/train.py", line 119, in train
> E           results = self._run_task(stage_module, mode="train", dataset=dataset, parameters=train_parameters)
> E         File "/home/cosmos/ws/otx-dev/otx/algorithms/common/tasks/training_base.py", line 168, in _run_task
> E           output = workflow.run(
> E         File "/home/cosmos/ws/otx-dev/otx/mpa/cls/trainer.py", line 38, in run
> E           cfg = self.configure(model_cfg, model_ckpt, data_cfg, training=True, **kwargs)
> E         File "/home/cosmos/ws/otx-dev/otx/mpa/cls/stage.py", line 28, in configure
> E           self.configure_model(cfg, model_cfg, training, **kwargs)
> E         File "/home/cosmos/ws/otx-dev/otx/mpa/cls/stage.py", line 37, in configure_model
> E           cfg.merge_from_dict(model_cfg._cfg_dict)
> E         File "/home/cosmos/anaconda3/envs/otx-dev/lib/python3.8/site-packages/mmcv/utils/config.py", line 644, in merge_from_dict
> E           Config._merge_a_into_b(
> E         File "/home/cosmos/anaconda3/envs/otx-dev/lib/python3.8/site-packages/mmcv/utils/config.py", line 327, in _merge_a_into_b
> E           b[k] = Config._merge_a_into_b(v, b[k], allow_list_keys)
> E         File "/home/cosmos/anaconda3/envs/otx-dev/lib/python3.8/site-packages/mmcv/utils/config.py", line 327, in _merge_a_into_b
> E           b[k] = Config._merge_a_into_b(v, b[k], allow_list_keys)
> E         File "/home/cosmos/anaconda3/envs/otx-dev/lib/python3.8/site-packages/mmcv/utils/config.py", line 327, in _merge_a_into_b
> E           b[k] = Config._merge_a_into_b(v, b[k], allow_list_keys)
> E         [Previous line repeated 1 more time]
> E         File "/home/cosmos/anaconda3/envs/otx-dev/lib/python3.8/site-packages/mmcv/utils/config.py", line 312, in _merge_a_into_b
> E           if allow_list_keys and k.isdigit() and isinstance(b, list):
> E       AttributeError: 'int' object has no attribute 'isdigit'
> ```

Overall, it's a workaround that looks confusing. In the light of the upcoming model API integration, we have to keep the inference-related code clean. I'd prefer keeping similar workarounds closer to config handling, rather than spreading the consequences of mmcv config limitations to the whole code of otx.
